linera-core/src/updater.rs

```diff
@@ -291,16 +293,38 @@ where
                     let certificates = local_storage
                         .read_certificates(blob_states.into_iter().map(|x| x.last_used_by))
                         .await?;
+                    let mut chain_heights = BTreeMap::new();
                     for certificate in certificates {
                         let block_chain_id = certificate.value().chain_id();
                         let block_height = certificate.value().height();
-                        self.send_chain_information(
-                            block_chain_id,
-                            block_height.try_add_one()?,
-                            CrossChainMessageDelivery::NonBlocking,
-                        )
-                        .await?;
+                        match chain_heights.entry(block_chain_id) {
+                            btree_map::Entry::Occupied(entry) => {
+                                let entry = entry.into_mut();
+                                if block_height > *entry {
+                                    *entry = block_height;
+                                }
+                            }
+                            btree_map::Entry::Vacant(entry) => {
+                                entry.insert(block_height);
+                            }
+                        }
                     }
+                    let stream = stream::iter(chain_heights)
+                        .map(|(chain_id, height)| {
+                            let mut client = self.clone();
+                            async move {
+                                let height = height.try_add_one()?;
+                                client
+                                    .send_chain_information(
+                                        chain_id,
+                                        height,
```

Comment:
Don't we need to add `1` here? (Or, alternatively, call `send_certificate` with the certificates we already read from storage above?)

Reply:
Not sure what you mean as there is an increase by 1 placed 4 lines above.

Reply:
Ah, thanks, I missed that! :facepalm: